### DISCUSSION TITLE
Dynamic FVM Calling Convention

### DISCUSSION THREAD

**Discussion by Stebalien** [filecoin-project/FIPs#409] [Enhancements - Technical]
This is an alternative proposal to #382.

Basically, we have two problems:

1. We need to be able to efficiently call methods from off-chain messages. In this case, the caller will likely know the actor's type ahead of time, and will be able to parse an IDL.
2. We need to be able to flexibly call methods from another actor on-chain. In this case, the caller likely won't be able to parse an IDL for performance reasons (costs to load the IDL, parse it, find the right method, etc.).

#382 proposes solving both problems by hashing the method name to a method number. However, this has some drawbacks:

- The receiver needs to know its set of methods ahead of time, limiting dynamic dispatch.
- The method needs to match _exactly_ (due to hashing).
- Debugging may be tricky as the FVM itself will only see the method number, not the actual method name.

# Proposal

An alternative is to break this problem into "dynamic" and "static" dispatch:

- Dynamic Dispatch: The caller sends a `(method_desc, params)` message to a reserved DYNAMIC_DISPATCH method number. This lets the caller tell the receiver what method the caller wants to invoke, and the receiver can then decide how it wants to handle the call (possibly delegating based on a namespace, possibly pattern matching based on some naming convention, etc.).
- Static Dispatch: The caller sends a message to a specific method number, on a specific actor type. The caller needs to know the type of actor it's invoking the method on for this to work. This lets a caller call a _specific_ method efficiently from off-chain.

```rust
struct DynamicCall {
    method: String,
    params: Box<CBOR>, // or by CID.
}
```

Note: The DYNAMIC_DISPATCH method number could just be `1`. This is currently used as the "constructor" method number, but that will likely be replaced with a separate actor entrypoint.

## Extension: Reflection

The basic proposal only sends a method name, but we could send a more complete "method description":

```rust
struct MethodDescription {
    name: String,
    namespace: Option<String>,
    signature: Option<??>,
}

struct DynamicCall {
    method: MethodDescription,
    params: Box<CBOR>, // or by CID.
}
```

This would let the receiver match on additional information, instead of _just_ matching on the method name.

# Analysis

The primary benefits of this approach are:

- This keeps method numbers, allowing actors to keep their existing ones as well, minimizing the required changes.
- Actors can prefer "small" method numbers for efficient messages (smaller numbers encode to smaller objects).
- On-chain, putting a method description into the method parameters doesn't add much overhead. It won't be stored on-chain, so the method name, namespace, etc. can be descriptive.
- Off-chain, users can chose to either use dynamic dispatch on-chain, or statically specify their method number (looking it up on docs, some IDL, etc.).

The primary drawbacks are:

- Runtime performance. The extra flexibility has a cost, but it should be minimal when compared to the rest of the decoding.
- Complexity.

**Comment by anorth** [comment]
I see this as complementary to #382. That is, adopt #382 as a base convention and then adopt this proposal as a new FRC for dynamic dispatch capabilities at method number `hash("Dynamic")`. It would be reasonable for actors to expose the same functionality behind each dispatch scheme (though tbh I don't see dynamic catching on widely). But this answers "The receiver needs to know its set of methods ahead of time, limiting dynamic dispatch" for receivers where that's actually an issue.

> Static Dispatch: The caller sends a message to a specific method number, on a specific actor type. The caller needs to know the type of actor it's invoking the method on for this to work. 
> ... This keeps method numbers, allowing actors to keep their existing ones as well, minimizing the required changes

I'm inferring from the above that your static proposal is _without_ the hashing described in #382. I think this would be really bad for composability and the development of standards. Widespread APIs like tokens need to be called from off-chain, efficiently, and using the _same_ method number across implementations. Needing to know the concrete type of the receiving actor is no good. We need a mechanism to coordinate non-colliding method numbers.

Note that #382 doesn't prevent actors from keeping existing method numbers, especially the built-in ones which are the only actors that could really be said to exist today (but see #401: most such methods should be hidden).

**Reply by Tom-OriginStorage** [reply]
Continuing the discussion from #382, I would like to reiterate the aspects which I disagree based on my personal experience (my opinion maybe myopic and not representative).
> but I don't think they'll hold when actor authors are optimising gas usage to provide a better (cheaper) product to their users.

I don't think **users** cares significantly about gas fees outside of Ethereum. We have Binance, Polygon, Harmony, Avax, Tron, Fantom, Cronos, Klatyn, Moonriver, Moonbeam, Fantom, Astar, Optimism, Arbitrum, Starknet, Zksync, Boba, Palm and many more chains running on EVM. Yet none of these chains managed to surpass Ethereum in terms of adoption in any aspect despite the cheaper gas fees. On the other hand, you have chains like Solana surpassing Ethereum in terms of NFT adoption (source from DappRadar trading volume & active wallets) despite not being an EVM chain. My point is that reducing the average message size by a few bytes would be the least of our concern. And moreover we are assuming here that following the Ethereum convention will consume less gas fees overall, which may not even be true due to how compilation works (although I must admit using Ethereum conventions is more likely to consume less gas fees).

And now let's take a look at the price we need to pay for trying to do this:
1. Block explorers will be displaying gibberish without the aid of an IDL/ABI for user-programmed actors. 
2. Data scrapers unable to scrape data without an IDL/ABI, and unable to build their applications around it. (when centralized entities refused to release their IDL/ABI)
3. Block explorers now need to integrate an upload their IDL/ABI feature.
4. Frontends now need to translate this IDL/ABI, someone needs to build & maintain this tool, Protocol Labs is already flooded with tasks. No need to if we are just using internal dynamic dispatch with json; just base64-cbor encode and you are free to go.
5. Closing ourselves to potential extension using the methods numbers such as read-only functions. While Protocol Labs has no intentions, I strongly believe developers planning to use Solidity for FVM development will need these read-only functions, and someone will implement it if the Filecoin ecosystem makes it.

Personally I think the cons outweigh the pros if we choose to adopt the Ethereum convention. Most builders out there don't even plan to use Rust to code on FVM (they are just waiting for Solidity). I think it would be better to save everyone the trouble and just use an internal dynamic dispatch which comes with some potential upside and compatibility with the current ecosystem.

> You'll have to give a more detailed idea of what you're describing here and what features of the original proposal lock it out.

This was an example of innovation Filecoin could potentially have. A pre-paid transaction. Users can pre-deposit an amount of crypto (which may not necessarily be FIL) to an account/actor. By specifying the method number, miners/validators are given the authority to deduct gas fees from that particular account/actor. The account/account sponsoring the transaction can be pooled and shared with other users to amortize gas costs overtime and even gain passive yield through staking. And Defi applications can use this sort of mechanisms to onboard users to new ecosystem.

**Reply by Stebalien** [reply]
> I'm inferring from the above that your static proposal is without the hashing described in https://github.com/filecoin-project/FIPs/discussions/382.

I am.

> Widespread APIs like tokens need to be called from off-chain, efficiently, and using the same method number across implementations. Needing to know the concrete type of the receiving actor is no good. We need a mechanism to coordinate non-colliding method numbers.

Why? If I'm calling an actor from off-chain, why can't I rely on knowing it's type ahead-of-time? I'd get a fair amount of additional safety if I check the type/method locally.

> Note that https://github.com/filecoin-project/FIPs/discussions/382 doesn't prevent actors from keeping existing method numbers, especially the built-in ones which are the only actors that could really be said to exist today (but see https://github.com/filecoin-project/FIPs/discussions/401: most such methods should be hidden).

With https://github.com/filecoin-project/FIPs/pull/434, I agree. At that point we can just support both and we can (by convention) reserve additional method-number ranges for additional calling conventions.

**Reply by Stebalien** [reply]
> Block explorers will be displaying gibberish without the aid of an IDL/ABI for user-programmed actors.

This is going to be an issue regardless. There's even a new working group in slack (#fvm-idl) to discuss this.

Unfortunately, if we start sending messages with full method names from off-chain, users _will_ optimize and start picking short/garbage method names. We really don't want that.

**Reply by anorth** [reply]
> > Widespread APIs like tokens need to be called from off-chain, efficiently, and using the same method number across implementations. Needing to know the concrete type of the receiving actor is no good. We need a mechanism to coordinate non-colliding method numbers.

> Why? If I'm calling an actor from off-chain, why can't I rely on knowing it's type ahead-of-time? I'd get a fair amount of additional safety if I check the type/method locally.

I need to explain better. "Off-chain" was a red herring, it's beside the point.

We need a mechanism to coordinate non-colliding method numbers for efficient composability. The motivating situation is that two actor authors who know nothing about each other need to coordinate. Consider someone implementing a fungible token, and another party implementing a DEX that swaps between tokens. Neither know about each other, but they both know about a standard API that assigns particular method numbers and types to methods for "Balance" and "TransferFrom". The DEX wants to be compatible with all tokens, including those that haven't been written yet. The token author wants to be compatible with all DEXs and other such tools, including those that haven't been written. That's why we need a standard. 

And we need coordination over the method number space, else two unrelated standards may collide on method numbers, making it impossible to implement both at once. With many standards, the probability of some mutual collision grows rapidly.

We could achieve this with string method names too. My proposal is merely a compression so as to avoid any bias against long method names (collision-resistance again). Your `DynamicCall` struct is actually my penultimate design for this, as in described https://github.com/filecoin-project/FIPs/discussions/382#discussioncomment-3152646

**Reply by Tom-OriginStorage** [reply]
> > Block explorers will be displaying gibberish without the aid of an IDL/ABI for user-programmed actors.
> 
> This is going to be an issue regardless. There's even a new working group in slack (#fvm-idl) to discuss this.
> 
> Unfortunately, if we start sending messages with full method names from off-chain, users _will_ optimize and start picking short/garbage method names. We really don't want that.

Unfortunately, I beg to differ. In reality, users will not start to optimize and pick short/garbage method names. This only happens to blockchains that can reach Ethereum level of adoption (Filecoin is nowhere near there). Cosmwasm is actually using this standard, but developers in the Cosmos ecosystem do spell out the names of the methods and gives zero damn about reducing the gas costs. I will need actual substantial evidence if you want to convince me that users will optimize by picking short/garbage method names.

**Reply by Stebalien** [reply]
> That's why we need a standard.

That's what I'm proposing here. The difference is that _this_ standard uses verbose method names.

> And we need coordination over the method number space, else two unrelated standards may collide on method numbers, making it impossible to implement both at once. With many standards, the probability of some mutual collision grows rapidly.

This assumes that the "standard" would rely on method numbers at all. The proposal here is to use method _names_ in the standard, and to have method-numbers be actor-specific.

> My proposal is merely a compression so as to avoid any bias against long method names (collision-resistance again).

This is where off-chain versus on-chain comes into play.

1. From off-chain, I can pre-resolve a method name to a method number that's specific to the actor I'm sending the message to. I can do this _off-chain_ because the operation may be expensive.
2. In messages between actors, per-byte gas is about 100x cheaper (and virtually free compared to the cost of a send).

Of course:

1. Even a tiny bit of gas might make a difference.
2. There may be cases where you need to specify the method from off-chain while not knowing the target actor ahead of time (some kind of late-bound call).

---

At this point, I'm just trying to make sure we fully understand each other. I'd be OK with a number-based calling convention (given some metadata on-chain for debugging). I'm just not convinced it's necessary (and it's definitely less flexible).

**Reply by anorth** [reply]
Ok, I think we're close to understanding each other. My summary of your proposal is that there are _two_ ways of invoking any method: one by name and one by number. The number depends on the concrete actor being invoked. Actors conventionally accept both dispatch methods (with some specific number meaning "dispatch by name" and a wrapped payload that carries that name). A Dapp or wallet wishing to use the cheaper by-number invocation will have to learn the method name→number mapping for the actor in question, perhaps by invoking a different method (by number?) that returns a table.

If I've got that right, then I think complexity (which you identified) is the most significant drawback. Two ways to do anything is just more complex. And here we're pushing some of that complexity out to Dapps and wallets (but the actors are a little more complex too).

I see rather little benefit in comparison to the method name hashing proposal for most cases. Method hashing fixes the name→number mapping globally for much less complexity. Rejection sampling means all existing method numbers can remain in use (and with plenty of room for more). Your proposal does offer ability to exploit smaller method numbers. I can also see potential utility in very dynamic cases, e.g. where the range of method names isn't even known at compile time but generated at internal dispatch (like implementing Python `__getattr__`).

If we wanted to make those very dynamic cases feel native from the start (rather than be behind a `hash("Dynamic")` method number), then dispatch by string from the get-go would be good. Honestly, I'd rather have just one way to do it and that way be by name, than have two ways. Since these are just recommendations to developers, we don't really have to choose for them, except that (a) they'll probably follow us, and (b) we should implement them in the built-in actors if we want them to play nicely.
